index.bs

```diff
@@ -346,10 +346,56 @@ interface mixin DatagramTransport {
 
-The <dfn>readDatagrams</dfn> algorithm is given a |transport| as parameter. It
-is defined by running the following steps:
-1. TODO
-1. [=ReadableStream/enqueue=] |data| in |transport|.`[[IncomingDatagrams]]`.
+To <dfn>pullDatagrams</dfn>, given a {{WebTransport}} object |transport|, run these steps:
+1. Assert: |transport|'s [=[[IncomingDatagramsPullPromise]]=] is null.
+1. Let |queue| be |transport|'s [=[[IncomingDatagramsQueue]]=].
+1. If |queue| is empty, then:
+  1. Let |promise| be a new promise.
+  1. Set |transport|'s [=[[IncomingDatagramsPullPromise]]=] to a new promise.
+  1. [=Upon fulfillment=] of |transport|'s [=[[IncomingDatagramsPullPromise]]=] with |bytes|, run
+     these steps:
+    1. Let |chunk| be a new {{Uint8Array}} object representing |bytes|.
+    1. [=ReadableStream/Enqueue=] |chunk| to |transport|'s [=[[Datagrams]]=]' s
+       [=DatagramDuplexStream/[[Readable]]=].
+    1. [=Resolve=] |promise| with undefined.
+  1. Return |promise|.
+1. Let |bytes| and |timestamp| be the result of [=dequeuing=] |queue|.
+1. Let |chunk| be a new {{Uint8Array}} object representing |bytes|.
+1. [=ReadableStream/Enqueue=] |chunk| to |transport|'s [=[[Datagrams]]=]' s
+   [=DatagramDuplexStream/[[Readable]]=].
+1. Return [=a promise resolved with=] undefined.
+
+To <dfn>receiveDatagrams</dfn>, given a {{WebTransport}} object |transport|, run these steps:
+1. Let |timestamp| be a timestamp representing now.
+1. Let |queue| be |transport|'s [=[[IncomingDatagramsQueue]]=].
+1. Let |duration| be |transport|'s [=[[IncomingDatagramsExpirationDuration]]=].
+1. If |duration| is null, then set |duration| to an [=implementation-defined=] value.
+1. Let |connection| be |transport|'s [=[[Connection]]=].
+1. While there are available incoming datagrams on |connection|: [[!WEB-TRANSPORT-HTTP3]]
+
+  Issue: We should probably use "session" instead of "connection", to support pooling.
+
+  1. Let |datagram| be the first datagram, and remove it from |connection|.
+  1. Let |timestamp| be a timestamp representing now.
+  1. Let |chunk| be a [=pair=] of |datagram| and |timestamp|.
+  1. [=Enqueue=] |chunk| to |queue|.
+1. Let |toBeRemoved| be the length of |queue| minus |transport|'s
+   [=[[IncomingDatagramsHighWaterMark]]=].
+1. If |toBeRemoved| is positive, repeat [=dequeuing=] |queue| |toBeRemoved| times.
+1. While |queue| is not empty:
+  1. Let |bytes| and |timestamp| be |queue|'s first element.
+  1. If more than |duration| milliseconds have passed since |timestamp|, then [=dequeue=] |queue|.
+  1. Otherwise, [=break=] this loop.
+1. If |queue| is not empty and |transport|'s [=[[IncomingDatagramsPullPromise]]=] is non-null, then
+   [=queue a global task=] on the [=network task source=] with |transport|'s
+   [=relevant global object=] and a task that runs the following steps:
+  1. Let |bytes| and |timestamp| be the result of [=dequeuing=] |queue|.
+  1. [=Resolve=] |transport|'s [=[[IncomingDatagramsPullPromise]]=] with |bytes|.
```

Comment:
In practice in Blink we can't resolve a promise with an abstract type, it has to be a JavaScript object. The obvious choice would be a Uint8Array, but that would be a security hole because the page might set `Object.prototype.then`. So we'd either use a custom object with a null prototype, or we'd replace the promise with a call to the relevant steps of _pullDatagrams_. The second option is cheaper and less complex.

As such, I'm wondering if we should instead write it as a call to another algorithm in the standard, instead of using an internal promise.

Reply:
Done.